### PR TITLE
Add asset handling

### PR DESCRIPTION
**Pull request by SamStenton**
Adds support for handling assets as they're not resolved by default

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 node_modules
 .DS_Store
+package-lock.json
```
Comment (mattdean1) [inline .gitignore]: If we stop tracking this file don't we also want to delete it from the repo?

File: test/resolve-test-data.js [hunk 2]
```diff
@@ -5,6 +5,41 @@ exports.raw = [
...
+      createdAt: `2017-09-13T14:15:08.218Z`,
+      updatedAt: `2017-10-02T09:49:40.790Z`,
+    },
+    fields: {
+      title: {
+        "en-US": `AWS S3`,
+      },
+      file: {
+        "en-US": {
+          url: `//images.contentful.com/im67laxslxvo/4jM3XYHBu0UACAIMeAgA86/87f6d5a43d86495abe84e4ec5012ad66/AWS_S3__1_.png`,
```
Comment (mattdean1) [inline test/resolve-test-data.js:23]: is this missing 'https://'?

**Comment by mattdean1** [comment]
Thanks for the PR! Extra 👍  for adding tests.

See above for a couple of queries. I reckon it might be worth adding a little something to the readme to explain what exactly happens to assets (is the file saved in redis?).

Looks like I need to set up CI properly for PRs 🔥

File: test/resolve-test-data.js (hunk 2 above)
Comment (SamStenton) [reply test/resolve-test-data.js:23]: Thats how it comes from Contentful I think

File: .gitignore (hunk 1 above)
Comment (SamStenton) [reply .gitignore]: 👍

**Comment by SamStenton** [comment]
[NPM](https://docs.npmjs.com/files/package-lock.json) says the lock file should be tracked. News to me!

**Comment by SamStenton** [comment]
Something else to think about with this pull is that calling `getResolvedEntries` will return entries with resolved assets but also the single asset objects. 

As a consumer of the module I'd want a separate `getAssets` method.

What are your thoughts? I've got an implementation of the above ready to go if you agree.

**Comment by mattdean1** [comment]
Yeah I'd reason as a module consumer it'd be nice to have 3 functions, `getEntries`, `getAssets`, and `getAll`, nice thinking.

Cheers for sorting the issues. Will it be too much work for you to split your assets stuff like this? I'm happy to get involved if needed.

**Comment by SamStenton** [comment]
That should do it! Anything else?

**Comment by mattdean1** [comment]
Last change! https://github.com/SamStenton/contentful-sync-redis/blob/582910bcb6bc51d828f2258129e5e9ca044a6e71/test/test.js#L8 Let's add the new functions to this test. 💪 

Might make this the big 1.0 release - thoughts?

**Comment by SamStenton** [comment]
Would you want the preview api to work for 1.0? Annoyingly it doesn't support sync so we'd have to change quite a bit. 

If we added support for the preview API it might also be worth looking at an option where redis is skipped entirely. This is essentially what would happen when using preview. Unless we relied on webhooks for updates.

**Comment by SamStenton** [comment]
To add to that. We recently had a convo with Contentful at work about best practices. Their recommendation was that for small apps, calling their API directly is better and has negligible speed impact. 

Possibly out of scope for 1.0 but would be cool to be able to switch redis/database storage on and off.

**Comment by mattdean1** [comment]
Sorry for slow reply 🥉 

My original plan for this package was to implement the sync API - I agree it'd be nice to have an option to turn sync on/off but I feel like that's out of scope here. I'd say with Preview API it'd be best to use the Contentful SDK directly, same with smaller apps as you mention above. 

I think the best solution would be **another package abstracting sync/preview/sdk functionality**, so users can make one function call in their code and change the data source via config. 

Additionally, another package '*contentful-sync-database*' could be created, abstracting the sync functionality and exposing a small interface to plug in an arbitrary database (Redis, *SQL, memcached?) ... at some point though we get diminishing returns since people can implement the Sync API themselves.

For now, I'm gonna merge this, and increment the minor version number then publish. Planning to improve test coverage then publish 1.0. 

Feel free to add more thoughts here or chat 😄

**Comment by SamStenton** [comment]
Thought that might be your answer! 😆  Sounds good to me. Give me a shout if there's anything I can do.

**Comment by nkhalidmms** [comment]
I'm looking forward to the merge to be completed and published as npm package. What is the timeline for this?

**Comment by SamStenton** [comment]
;)

**Pull request closed without merging**